pkg/controller/elasticsearch/sset/pod.go

```diff
@@ -76,29 +77,57 @@ func GetActualMastersForCluster(c k8s.Client, es v1alpha1.Elasticsearch) ([]core
 	return label.FilterMasterNodePods(pods), nil
 }
 
-// ScheduledUpgradesDone returns true if all pods scheduled for upgrade have been upgraded.
-// This is done by checking the revision of pods whose ordinal is higher or equal than the StatefulSet
-// rollingUpdate.Partition index.
-func ScheduledUpgradesDone(c k8s.Client, statefulSets StatefulSetList) (bool, error) {
-	for _, s := range statefulSets {
-		if s.Status.UpdateRevision == "" {
-			// no upgrade scheduled
-			continue
+func PodReconciliationDoneForSset(c k8s.Client, statefulSet appsv1.StatefulSet) (bool, error) {
+	// check all expected pods are there: no more, no less
+	actualPods, err := GetActualPodsForStatefulSet(c, statefulSet)
+	if err != nil {
+		return false, err
+	}
+	actualPodNames := k8s.PodNames(actualPods)
+	expectedPodNames := PodNames(statefulSet)
+	if !(len(actualPodNames) == len(expectedPodNames) && stringsutil.StringsInSlice(expectedPodNames, actualPodNames)) {
+		log.V(1).Info(
+			"Some pods still need to be created/deleted",
+			"namespace", statefulSet.Namespace, "statefulset_name", statefulSet.Name,
+			"expected_pods", expectedPodNames, "actual_pods", actualPodNames,
+		)
+		return false, nil
+	}
+
+	// check pods revision match what is specified in the StatefulSet
+	done, err := scheduledUpgradeDone(c, statefulSet)
+	if err != nil {
+		return false, err
+	}
+	if !done {
+		log.V(1).Info(
+			"Some pods still need to be upgraded",
+			"namespace", statefulSet.Namespace, "statefulset_name", statefulSet.Name,
+		)
+		return false, nil
+	}
+
+	return true, nil
+}
```

Comment:
When moving to `onDelete`, I would expect this to become:

1. check that the expected number of pods is running (same first half of the function)
2. check our deletions expectations: if we did `client.Delete(pod, ...)` at some point, we should make sure this pod has indeed disappeared from our cache here. This would replace the second half of the function: we would not really care about revision here. As long as all pods are there, and no deletion is in progress (for potential pod recreation by the sset controller), I think we should be fine?